lib/common/ipc_client.c

```diff
@@ -510,16 +511,32 @@ pcmk_connect_ipc(pcmk_ipc_api_t *api, enum pcmk_ipc_dispatch dispatch_type)
     }
 
     api->dispatch_type = dispatch_type;
-    switch (dispatch_type) {
-        case pcmk_ipc_dispatch_main:
-            rc = connect_with_main_loop(api);
-            break;
 
-        case pcmk_ipc_dispatch_sync:
-        case pcmk_ipc_dispatch_poll:
-            rc = connect_without_main_loop(api);
+    for (int i = 0; i < n_attempts; i++) {
+        switch (dispatch_type) {
+            case pcmk_ipc_dispatch_main:
+                rc = connect_with_main_loop(api);
+                break;
+
+            case pcmk_ipc_dispatch_sync:
+            case pcmk_ipc_dispatch_poll:
+                rc = connect_without_main_loop(api);
+                break;
+        }
+
+        if (rc != EAGAIN) {
             break;
+        }
+
+        /* EAGAIN may occur due to interruption by a signal or due to some
+         * transient issue. Try one more time to be more resilient.
+         */
+        if (i < (n_attempts - 1)) {
```

Comment:
I plan to keep the condition in the loop header for clarity, but if we're essentially checking for termination here then the loop condition is unnecessary.